File: fs-regiondeployer/logger.fs
namespace regiondeployer

open System
open System.IO
open Microsoft.FSharp.Core

type private logginglock() =
    static member val public lock = new Object()
        with get

let private logfile : string = (Directory.GetCurrentDirectory() + "/log.txt")

let public initialize() : unit = 
    use init = File.Create(logfile)
    ()

///logs to console and local file
///does not work in a unit test
let public logtoconsoleandfile (message:string) : unit =
    lock logginglock.lock (fun _ -> 
        Console.WriteLine message
        use logfilestream = File.AppendText(logfile)
        logfilestream.WriteLine(message)
    )
    

File: regiondeployer/Logger.cs
using System;
using System.IO;

namespace regiondeployer
{
    public class LoggingLock
    {
        public static readonly object Lock = new object();
    }

    public class Logger
    {
        private readonly string _logFile;

        public Logger()
        {
            _logFile = Directory.GetCurrentDirectory() + "/log.txt";
            using (var init = File.Create(_logFile)) { }
        }

        public void LogToConsoleAndFile(string message)
        {
            lock (LoggingLock.Lock)
            {
                Console.WriteLine(message);
                using (var logFileStream = File.AppendText(_logFile))
                {
                    logFileStream.WriteLine(message);
                }
            }
        }
    }
}